decisions/DR006-Generic-string-type.md

```diff
@@ -0,0 +1,129 @@
+# DR006 Generic string type
+
+- **Status:** Proposed
+- **Impact:** Medium
+- **Driver:** @feltech
+- **Approver:** @foundrytom
+- **Contributors:** @feltech @foundrytom
+- **Outcome:** TBD
+
+## Background
+
+We must decide what generic string type should be exposed in the
+API(s).
+
+We must consider any potential ABI issues as well as the C APIs that
+rely on strings.
+
+## Relevant data
+
+Originating [GitHub issue](https://github.com/TheFoundryVisionmongers/OpenAssetIO/issues/193).
+
+Assumptions:
+* UTF-8 encoding is used throughout, wherever there is a choice. Any 
+  necessary conversions (e.g. for Windows UIs/files) is the 
+  responsibility of the host/plugin.
+* Python is restricted to Python 3, where we use the native `str` type,
+  which is unicode but [not necessarily](https://docs.python.org/3/c-api/unicode.html)
+  UTF-8. However, we assume C++ bindings via [pybind11](https://pybind11.readthedocs.io/en/stable/advanced/cast/strings.html),
+  which encodes Python->C++ strings as UTF-8 (`char*` or `std::string`)
+  and expects pre-encoded UTF-8 strings for C++->Python.
+* Some strings should be strongly typed, e.g. entity references, but we
+  still require a generic string type for some purposes.
+
+References:
+* [GCC C++11 ABI break](https://gcc.gnu.org/onlinedocs/libstdc++/manual/using_dual_abi.html)
+  ([VFX reference platform](https://vfxplatform.com/) is currently stuck
+  on the old ABI).
+* [UTF8-CPP](https://github.com/nemtrif/utfcpp) C++ library provides 
+  functions for operating on UTF-8 strings stored within a 
+  `std::string`.
+* IBM's [ICU](https://icu.unicode.org/) popular heavyweight C (unicode)
+  string library with custom types.
+* [CsString](https://github.com/copperspice/cs_string) library providing
+  dedicated unicode-aware string types.
+* [ACCU Overload, 21(116):11-15, 2013](https://accu.org/journals/overload/21/116/steinbach_1842/)
+  article on portable string literals.
+* [utf8everywhere](http://utf8everywhere.org/) article promoting "just 
+  use UTF-8".
+
+## Options considered
+
+### Option 1
+
+* C++: `std::string` for constants/parameters.
+* C: `char*` access/assignment.
```

Comment:
Option 1a:

C: `struct strview { char* start, char* end }`

C char* strings have the burden of needing to measure a string in order to work with it, and the burden of needing a zero terminator. A strview with a guarantee, at least from the library, that it was constructed with an end > start, and a pair of pointers that actually span a string is in a much better spot for language bindings and other operations. Additionally, a C++ conversion operator can intern the strview for std::string.

std::string_view for C++ might also be a good option, if the strings are interned by the library, and can be treated as immutable. That usage would shift the burden of memory copies to the user of the library, rather than baking it into the API.

Reply:
Ah yes, my mind as still stuck in the dark ages of C++11/14.  `string_view` for constants makes a lot of sense (and can be `constexpr` in C++).

Intriguing idea to create C `string_view`-like type.  Is this something you have seen used in other projects?

Reply:
Little [Compiler Explorer](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGISQDspK4AMngMmAByPgBGmMT%2BABykAA6oCoRODB7evv5BaRmOAmER0SxxCZLJdpgOWUIETMQEOT5%2BgbaY9sUMjc0EpVGx8Um2TS1teZ0KE4PhwxWj1QCUtqhexMjsHOYAzOHI3lgA1CZ7brPE4cAA%2BgBueJgA7ufYJhoAgvuHx5hnFycV0wrDeH2%2BZgODCOXlO5zcyCuNzBXx%2B0L%2BAIRTAUShaKO%2BXzQDFmmFUKWIJ1m6BAICRRgeT2eJwA1gCACJnMxmBSoNiUgjXIzmMznKyor6kgjxBicsxuYVnVEBMUEz5XLwOE6IgU3BkvcEmZXgk4mrUIZpagSzE4AKktxIIJziwHCouNpuQ5opROtdp9jtcbqVbKDqvB4UdwEwBFu2uIEDjuseL3t1uTz1IZotfqtjo2BEzVJpGQAXphbo7/ScWExVPQGCsDUavqaTtjcQQIDW64wAdgThpG3sVe6TQB6McnIR4MucixcE78CkMLy0WgnKXEFjhJhEYgAOlHqcdTC8REpJfZ/N3eGQsexBHhRdpM/Lj722Ag6f3rhOAFoTm/Z1whWOcTi4UNWyrU9zyYe4mDoJgYnoW4FEvc4OWfbcGAgNDM27eshxHFtTWfK5oRSABPCB80zIDMBdBhMzghCDGQ8s0KIg0SJNfMTAAVgsFjEPY1DLwAiD%2BI5DCTjAXZ%2BLcDQ5NDI9iGjTYZWEtiULQlSAhDYMTgnE5JWlWV5S5bjVQjasEJwmzmmAZBM09ZobTtRz7kbJUVVbYyPAdYgNUdNwTmeYgmBSFJ4hOVAqGPQxHTpO50yPRN6XTLVmUNCxmX3YCcLWFkfwYdAIEbfSVJ4ozJwC%2B54hCsKIqimKYiYZBWXPCyrEsflBRSxlD2qmDUGPC8r1mG871EWYnwIalaR1DLGWLV8K0o6K3gTPLfwAjr8oYkDINI%2BbiyWgaUyrZlsuVfaCsLEtDRDYcrL82rUCoxdV3XULMq6rMKWaCLKKGz4oK9L611jK4BPnDQNAE57fNNKMYzjbbMyob7oYFQtXzi6hsbjFYuPFMHTX8j7KLbSlaFvf44pOX7GQ3UbQtcwHiGB0HwYtFIBjwMQceIWHJER46TVR4WMZOfmWkF2hhbxssCblxwheJ0nyZNV6ToWtAz0xeFZVZGTlK5I2LhZS3QuFZThxZU29g5c2zBtx33aPVs7d2B2sahuMr1d93/cVwP4WNr3TR90VZYFjWriD3YLYjq21YV4XPa5YPU9C59XFoKrtZONSCA04rS0wcr/0d/dK/KvS2Q4NZaE4fjeD8DgtFIVBOG66xKQ2LZ/n2HhSAITRm7WZl/C4fdJDMKR%2BICAA2AI9i4fj%2BLMAJpFbjhJA7yee84XgFBADRx8ntY4FgGBEBQXkUjoeJyEoNAWGf%2BgEiOQxgACeGfA6CbnPhAGIx8Yg7mIJRTgY8P5sEEAAeQYLQGBXdeBYBrEYcQ6DSD4DUvUeq59cGkjqGeHY3cIzdGPnTGIEVoEeCwLA3gOoWDMLWFQAwwAFAADVGSIOip3Me/BBAiDEOwKQMhBCKBUOoXBuguD6D/igaw1h9B4BiOfWAzA2AgBiKgdIBZ6oJG4AAqePcUi9GIX%2BKkGFTCWGsFwAIZ9uh1F6C4UqUw/CKNCAscolQ9CFEyAILxgT0jBIYEMfxoxFG1HqAIfokxPDtD0HE3oiT5hlBGAkWJcxQm5IGFE7JEg1g8k2NsPQApMA7B4C3NuR9cG9w4KoRIK8/wr0kFqAwRgTgBH3BofpJwID90sJmXAhASCcg3pmDwn8X4UlHisFh19p7%2BD2PuMwK8NABEXsvAAnHsDQ/EV78X0JwQ%2BpA2H8Uvp3buTSz4Xyvugm%2B98IBIHzCkM8b8ExP3mXoTA%2BA9x6BEcIUQ4hJEgpkWoY%2B4VIrsLORwdupBbm8CaYgs8ny8zxRaW0jpXS/69P6YM4Zvzv5TK4Esp5WgSakAQCCLACRyoIouVcm5x97m2EeRPZ5qz%2BKJH3MvRI/EF57E2YkSQgQzAIr2A0u5p8qXmP3mYWVqL5XcupWsYxGRnCSCAA%3D) experiment of a C `string_view`.

Reply:
Added `string_view` based options in b80f4bf

Reply:
I've used that pattern in my own code for a couple of decades... Nowadays, I suppose this library might be called a parser combinator or something. https://github.com/meshula/LabText/blob/main/include/LabText/LabText.h 

The redis string library is an interesting similar example, although it's ptr+sz, rather than start+end. 
https://github.com/antirez/sds